src/storage/mvcc/consistency_check.rs

```diff
@@ -0,0 +1,462 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Ordering;
+use std::convert::TryInto;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicU64, Ordering as AtomicOrdering};
+use std::sync::Arc;
+
+use crate::storage::mvcc::{Lock, LockType, WriteRef, WriteType};
+use engine_traits::{
+    IterOptions, Iterable, Iterator as EngineIterator, KvEngine, Peekable, SeekKey,
+};
+use engine_traits::{CF_DEFAULT, CF_LOCK, CF_RAFT, CF_WRITE};
+use kvproto::kvrpcpb::{MvccInfo, MvccLock, MvccValue, MvccWrite, Op};
+use raftstore::coprocessor::{ConsistencyCheckMethod, ConsistencyCheckObserver, Coprocessor};
+use raftstore::Result;
+use tikv_util::keybuilder::KeyBuilder;
+use txn_types::Key;
+
+const PHYSICAL_SHIFT_BITS: usize = 18;
+const SAFE_POINT_WINDOW: usize = 120;
+
+// When leader broadcasts a ComputeHash command to followers, it's possible that the safe point
+// becomes stale when the command reaches followers. So use a 2 minutes window to reduce this.
+fn get_safe_point_for_check(mut safe_point: u64) -> u64 {
+    safe_point >>= PHYSICAL_SHIFT_BITS;
+    safe_point += (SAFE_POINT_WINDOW * 1000) as u64; // 120s * 1000ms/s.
+    safe_point << PHYSICAL_SHIFT_BITS
+}
+
+#[derive(Clone)]
+pub struct Mvcc<E: KvEngine> {
+    _engine: PhantomData<E>,
+    local_safe_point: Arc<AtomicU64>,
+}
+
+impl<E: KvEngine> Coprocessor for Mvcc<E> {}
+
+impl<E: KvEngine> Mvcc<E> {
+    pub fn new(safe_point: Arc<AtomicU64>) -> Self {
+        Mvcc {
+            _engine: Default::default(),
+            local_safe_point: safe_point,
+        }
+    }
+}
+
+impl<E: KvEngine> ConsistencyCheckObserver<E> for Mvcc<E> {
+    fn update_context(&self, context: &mut Vec<u8>) -> bool {
+        context.push(ConsistencyCheckMethod::Mvcc as u8);
+        context.reserve(8);
```

Comment:
Should it be `context.reserve(9);`?

Reply:
`8` is enough, because the type byte has already been writiten.